internal/states/encryption/encryption.go

```diff
@@ -0,0 +1,271 @@
+package encryption
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/go-hclog"
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionconfig"
+	"github.com/opentofu/opentofu/internal/states/encryption/encryptionflow"
+	"github.com/opentofu/opentofu/internal/tfdiags"
+)
+
+// New creates a new Encryption object. You can use this object to feed in encryption configuration and then create
+// an encryptionflow.StateFlow or encryptionflow.PlanFlow as needed.
+//
+// Note: a large portion of the OpenTofu codebase is still procedural, which means there is no way to properly inject
+// the Encryption object and carry the information it holds across subsystem boundaries. In most cases you should use
+// GetSingleton() to get a globally scoped copy of the Encryption object. However, for tests you should use this
+// function and hopefully, some time in the future, we can get rid of the singleton entirely.
+func New(logger hclog.Logger) Encryption {
+	return &encryption{
+		encryptionConfigs:         make(encryptionconfig.ConfigMap),
+		decryptionFallbackConfigs: make(encryptionconfig.ConfigMap),
+		mutex:                     sync.Mutex{},
+		logger:                    logger,
+	}
+}
+
+// Encryption is the main interface to feed the encryption configuration and obtain the encryptionflow.Flow for running
+// the actual encryption.
+//
+// Note: a large portion of the OpenTofu codebase is still procedural, which means there is no way to properly inject
+// the Encryption object and carry the information it holds across subsystem boundaries. In most cases you should use
+// GetSingleton() to get a globally scoped copy of the Encryption object. However, for tests you should use the New()
+// function and hopefully, some time in the future, we can get rid of the singleton entirely.
+type Encryption interface {
+	// ApplyEnvConfigurations applies a configuration coming from the operating system environment.
+	ApplyEnvConfigurations(
+		encryption map[encryptionconfig.Key]encryptionconfig.Config,
+		decryptionFallback map[encryptionconfig.Key]encryptionconfig.Config,
+	) error
+
+	// ApplyHCLEncryptionConfiguration applies a single encryption configuration coming from HCL.
+	ApplyHCLEncryptionConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error
+	// ApplyHCLDecryptionFallbackConfiguration applies a single decryption fallback configuration coming from HCL.
+	ApplyHCLDecryptionFallbackConfiguration(key encryptionconfig.Key, config encryptionconfig.Config) error
+
+	// Validate ensures that the previously-applied configuration is valid. You can call this function after you
+	// applied all configurations in order to verify that all configurations are valid.
```

Comment:
```suggestion
	// Validate ensures that the previously-applied configuration is valid. You should call this function after you
	// applied all configurations in order to verify that all configurations are valid.
```

Reply:
Happy to add this change, but you'll get the same error eventually if you call functions like `Statefile`. This only makes sure you check all possible configurations all at once.